src/requiredBy.js

```diff
@@ -0,0 +1,12 @@
+export default function requiredBy(requiredByPropName, propType, defaultValue = null) {
+  return function validator(props, propName, componentName, ...rest) {
+    if (props[requiredByPropName]) {
+      if (props[propName] === defaultValue || typeof props[propName] === 'undefined') {
+        return new TypeError(
+          `${componentName}: when ${requiredByPropName} is true, prop “${propName}” must be present.`,
+        );
+      }
+    }
+    return propType(props, propName, componentName, ...rest);
+  };
```

Comment:
we'll also want to make sure the validator has a `.isRequired` on it - basically, it would check if `props[propName] == null`, and if so, error out - if not, it would call into the non-required logic.

After that, there's a `wrapValidator` helper that should ideally be used here.